features/netsocket/emac-drivers/TARGET_RDA_EMAC/rda5981x_emac.cpp

```diff
@@ -0,0 +1,302 @@
+            * pbuf is the sum of the chained pbuf len members.
+            */
+            /* load rx data from 96 to local mem_pool */
+            memcpy(memory_manager->get_ptr(q), &data[index], memory_manager->get_len(q));
+            index += memory_manager->get_len(q);
+
+            if(index >= len)
+            {
+                break;
+            }
+        }
+
+    }
+    else
+    {
+        /* Drop this packet */
+        LWIP_DEBUGF(NETIF_DEBUG, ("low_level_input pbuf_alloc fail, rxlen:%d\n", len));
+
+        return NULL;
+    }
+    return p;
+}
+
+
+/** \brief  Attempt to read a packet from the EMAC interface.
+ *
+ */
+void RDA5981x_EMAC::packet_rx()
+{
+    rda_msg msg;
+    packet_rx_queue = rda_mail_create(10, sizeof(unsigned int)*4);
+    /* move received packet into a new buf */
+    while (1) {
+        emac_mem_buf_t *p = NULL;
+        rda_mail_get(packet_rx_queue, (void*)&msg, osWaitForever);
+        switch(msg.type)
+        {       
+            case 0:
+                p = low_level_input((unsigned char*)msg.arg1, msg.arg2);
+                if (p == NULL) {
+                    rda_sem_release((void*)msg.arg3);
+                    break;
+                }
+                rda_sem_release((void*)msg.arg3);
+                if (p) {
+                    emac_link_input_cb(p);
+                }
+                break;
+            case 1:
+                emac_link_state_cb(msg.arg1);
+                break;
+            default:
+                break;
+        }
+    }
+}
+
+void RDA5981x_EMAC::thread_function(void *pvParameters)
+{
+    static struct RDA5981x_EMAC *rda5981x_enet = static_cast<RDA5981x_EMAC *>(pvParameters);
+    rda5981x_enet->packet_rx();
+}
+
+bool RDA5981x_EMAC::power_up()
+{
+    /* Initialize the hardware */
+    static int init_flag = 0;
+    if(init_flag == 0){
+        wland_reg_func();
+        rda_thread_new("maclib_thread", maclib_task, NULL, DEFAULT_THREAD_STACKSIZE*8, PHY_PRIORITY);
+        rda_thread_new("wland_thread", wland_task, NULL, DEFAULT_THREAD_STACKSIZE*5, PHY_PRIORITY);
+        rda_thread_new("packet_rx", RDA5981x_EMAC::thread_function, this, DEFAULT_THREAD_STACKSIZE*5, PHY_PRIORITY);
+        /* Allow the PHY task to detect the initial link state and set up the proper flags */
+        osDelay(100);
+        wland_sta_init();
+        init_flag = 1;
+    }
+
+    return true;
+}
+
+uint32_t RDA5981x_EMAC::get_mtu_size() const
+{
+    return RDA_ETH_MTU_SIZE;
+}
+
+uint32_t RDA5981x_EMAC::get_align_preference() const
+{
+    return 0;
+}
+
+void RDA5981x_EMAC::get_ifname(char *name, uint8_t size) const
+{
+    memcpy(name, RDA_ETH_IF_NAME, (size < sizeof(RDA_ETH_IF_NAME)) ? size : sizeof(RDA_ETH_IF_NAME));
+}
+
+uint8_t RDA5981x_EMAC::get_hwaddr_size() const
+{
+    return RDA_HWADDR_SIZE;
+}
+
+bool RDA5981x_EMAC::get_hwaddr(uint8_t *addr) const
+{
+    mbed_mac_address((char *)addr);
```

Comment:
I don't see this function used anywhere else in the driver, are you sure that it actually provides the MAC address that this hardware uses?

If you don't provide implementation of `mbed_mac_address()` it gives you pseudo MAC address that is generated in build time.